FLEDGE.md

```diff
@@ -211,6 +212,8 @@ The returned `auctionResultPromise` object is _opaque_: it is not possible for a
 
 Optionally, `sellerTimeout` can be specified to restrict the runtime (in milliseconds) of the seller's `scoreAd()` script, and `perBuyerTimeouts` can be specified to restrict the runtime (in milliseconds) of particular buyer's `generateBid()` scripts. If no value is specified for the seller or a particular buyer, a default timeout of 50 ms will be selected. Any timeout higher than 500 ms will be clamped to 500 ms. A key of `'*'` in `perBuyerTimeouts` is used to change the default of unspecified buyers.
 
+Optionally, the `abortSignal` field can be set to an [AbortSignal](https://dom.spec.whatwg.org/#interface-AbortSignal) object to permit aborting the execution of the auction.  When the `abort()` method on the associated `AbortController` is called, an attempt to interrupt the auction will be made. Since the auction executes in parallel to the page, it's possible for this call to happen after the auction actually completed but before this has been noticed by the caller of `runAdAuction`. In that case, the cancellation attempt is ignored. If the cancellation is successful, the promise returned from `runAdAuction` will resolve to null.
```

Comment:
I believe this functionality may help address an issue brought up in https://github.com/WICG/turtledove/issues/318. That said, do we need to define the behavior around what happens with reporting (`reportWin` and `reportResult`) in the event when an auction is cancelled more explicitly? For example, if the cancellation is successful, would that signal a strong guarantee that reporting from the auction did not yet have a chance to happen?

Reply:
My understanding is that the reporting is done in the final step of runAdAuction() after the actual auction is completed. If the cancellation attempt is sent during the reporting period, it will be ignored. This is to guarantee no partial/conflict reporting being sent. Is this correct?

Reply:
Mostly. It indeed either it finishes, or reportWin/reportResult don't run. 

I think some of the pending aggregation reports may still run, though.  Perhaps they shouldn't? 
Also some of the statistics on number of bids, etc., will get updated.  That may be trickier...

Reply:
So I got some pushback on the possibility of abort failing on the CL that implements this.  In the alternative, the promise will always (except if it already resolved?) be rejected, and all the work thrown out (this direction is conditional on the change to delay reporting till after the auction completes).  What do you folks think of that?

(https://chromium-review.googlesource.com/c/chromium/src/+/3805648 if you folks want to look in, though the conversation is kinda scattered).

Reply:
Will this result in a race condition that is the promise gets rejected but the report is still sent? Imagine the cancellation request comes in during the reporting period.

Reply:
This would be contingent on changing the reporting to happen after the auction completes (perhaps when the ad actually loads) ---- note that I can't commit to us doing that, just say that a colleague is looking into that --- see the conversation here: https://chromium-review.googlesource.com/c/chromium/src/+/3805648/15/third_party/blink/renderer/modules/ad_auction/navigator_auction.cc#1620

Reply:
Then it won't be consistent with the comment 4 Paul made "If the auction is canceled, no reports are sent." and it won't fix the duplicate ping issue in #318. However, it is still valuable in terms of saving computational resources.

By "changing the reporting to happen after the auction completes", are you referring to delay sending report until the resulting URN is handed to a FF for render? as mentioned in the [spec](https://github.com/WICG/turtledove/blob/main/FLEDGE.md#5-event-level-reporting-for-now) and option 5 in Paul's [comment] (https://github.com/WICG/turtledove/issues/318#issuecomment-1176262550). Is there an issue to track the progress?

Reply:
So current status is that if it's aborted successfully/promise is rejected no side effects happen --- no reporting, no updates of bid counts, no private aggregation requests, etc.   If abort is too late and promise resolves, then it's a successful auction, with reporting and all.

Reply:
Thanks for the update Maks!

Reply:
We are looking into using this feature. One thing I am not sure if it is implemented consistently is that when the auction is cancelled with AbortSignal, the promise should be rejected with error set to [AbortSignal.reason](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal/reason)  which can be TimeoutError (if AbortSignal.timeout() is used) or an AbortError. In this way, we are able to differentiate a real error vs. timeout.

[1] https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal/timeout#return_value

Reply:
By looking at the code change, https://chromium-review.googlesource.com/c/chromium/src/+/3805648/21/third_party/blink/renderer/modules/ad_auction/navigator_auction.cc#1627

I believe it is implemented in this way.  Just want to confirm.

Reply:
It does appear to work fine (at least from looking at the serialization in .catch()); there is also some test coverage of passing of reason via AbortController.signal:

https://chromium-review.googlesource.com/c/chromium/src/+/3805648/21/content/browser/interest_group/interest_group_browsertest.cc#3432

https://chromium-review.googlesource.com/c/chromium/src/+/3805648/21/content/browser/interest_group/interest_group_browsertest.cc#3542